File: crypto/presentation/src/main/java/dev/ohoussein/crypto/presentation/reducer/CryptoListReducer.kt
package dev.ohoussein.crypto.presentation.reducer

import dev.ohoussein.crypto.presentation.model.Crypto
import dev.ohoussein.cryptoapp.common.mvi.Reducer
import dev.ohoussein.cryptoapp.common.mvi.UiEvent
import dev.ohoussein.cryptoapp.common.mvi.UiIntent
import dev.ohoussein.cryptoapp.common.mvi.UiState
import dev.ohoussein.cryptoapp.common.resource.DataStatus

class CryptoListReducer : Reducer<CryptoListState, CryptoListEvents>(CryptoListState()) {

    override fun reduce(oldState: CryptoListState, event: CryptoListEvents): CryptoListState {
        return when (event) {
            is CryptoListEvents.UpdateCryptoList -> {
                oldState.copy(cryptoList = event.cryptoList, status = DataStatus.Success)
            }
            is CryptoListEvents.UpdateStatus -> {
                oldState.copy(status = event.status)
            }
        }
    }
}

sealed class CryptoListEvents : UiEvent {
    data class UpdateCryptoList(val cryptoList: List<Crypto>) : CryptoListEvents()
    data class UpdateStatus(val status: DataStatus) : CryptoListEvents()
}

sealed class CryptoListIntent : UiIntent {
    object ScreenOpened : CryptoListIntent()
    object Refresh : CryptoListIntent()
}

data class CryptoListState(
    val cryptoList: List<Crypto>? = null,
    val status: DataStatus = DataStatus.Initial,
) : UiState

File: CryptoAppiOS/CyptoAppiOS/Crypto/Reducer/CryptoListReducer.swift
import Foundation

func cryptoListReducer(state: CryptoListState, event: CryptoListEvent) -> CryptoListState {
    var newState = state
    switch event {
    case let .updateCryptoList(list):
        newState.cryptoList = list
        newState.status = LoadingStatus.idle
    case let .updateStatus(status):
        newState.status = status
    }
    print("New status $\(newState.status)")
    return newState
}

struct CryptoListState {
    var cryptoList: [Crypto]?
    var status: LoadingStatus = .idle
}

enum CryptoListIntent {
    case refresh
    case hideError
}

enum CryptoListEvent {
    case updateCryptoList([Crypto])
    case updateStatus(LoadingStatus)
}
